iot/api-client/end_to_end_example/cloudiot_pubsub_example_mqtt_device.py

```diff
@@ -0,0 +1,242 @@
+
+
+def error_str(rc):
+  """Convert a Paho error to a human readable string."""
+  return '{}: {}'.format(rc, mqtt.error_string(rc))
+
+
+class Device(object):
+  """Represents the state of a single device."""
+
+  def __init__(self):
+    self.temperature = 0
+    self.fan_on = False
+    self.connected = False
+
+  def update_sensor_data(self):
+    """Pretend to read the device's sensor data.
+
+    If the fan is on, assume the temperature decreased one degree,
+    otherwise assume that it increased one degree.
+    """
+    if self.fan_on:
+      self.temperature -= 1
+    else:
+      self.temperature += 1
+
+  def wait_for_connection(self, timeout):
+    """Wait for the device to become connected."""
+    total_time = 0
+    while not self.connected and total_time < timeout:
+      time.sleep(1)
+      total_time += 1
+
+    if not self.connected:
+      raise RuntimeError('Could not connect to MQTT bridge.')
+
+  def on_connect(self, unused_client, unused_userdata, unused_flags, rc):
+    """Callback for when a device connects."""
+    print 'Connection Result:', error_str(rc)
+    self.connected = True
+
+  def on_disconnect(self, unused_client, unused_userdata, rc):
+    """Callback for when a device disconnects."""
+    print 'Disconnected:', error_str(rc)
+    self.connected = False
+
+  def on_publish(self, unused_client, unused_userdata, unused_mid):
+    """Callback when the device receives a PUBACK from the MQTT bridge."""
+    print 'Published message acked.'
+
+  def on_subscribe(self, unused_client, unused_userdata, unused_mid,
+                   granted_qos):
+    """Callback when the device receives a SUBACK from the MQTT bridge."""
+    print 'Subscribed: ', granted_qos
+    if granted_qos[0] == 128:
+      print 'Subscription failed.'
+
+  def on_message(self, unused_client, unused_userdata, message):
+    """Callback when the device receives a message on a subscription."""
+    payload = str(message.payload)
+    print "Received message '{}' on topic '{}' with Qos {}".format(
+        payload, message.topic, str(message.qos))
+
+    # The device will receive its latest config when it subscribes to the config
+    # topic. If there is no configuration for the device, the device will
+    # receive an config with an empty payload.
+    if not payload:
+      return
+
+    # The config is passed in the payload of the message. In this example, the
+    # server sends a serialized JSON string.
+    data = json.loads(payload)
+    if data['fan_on'] != self.fan_on:
+      # If we're changing the state of the fan, print a message and update our
+      # internal state.
+      self.fan_on = data['fan_on']
+      if self.fan_on:
+        print 'Fan turned on.'
+      else:
+        print 'Fan turned off.'
+
+
+def parse_command_line_args():
+  """Parse command line arguments."""
+  parser = argparse.ArgumentParser(
+      description='Example Google Cloud IoT MQTT device connection code.')
+  parser.add_argument(
```

Comment:
Maybe default some of these as described for the server:

* GCLOUD_PROJECT or GOOGLE_CLOUD_PROJECT - used for the project ID
* GOOGLE_API_KEY - used for API key (although the sample should not require it)
* GOOGLE_APPLICATION_CREDENTIALS - path to service.json

Reply:
Done.